src/apps/lyra-avalon/optimism/lyra-avalon.balance-fetcher.ts

```diff
@@ -0,0 +1,119 @@
+
+const appId = LYRA_AVALON_DEFINITION.id;
+const network = Network.OPTIMISM_MAINNET;
+
+@Register.BalanceFetcher(LYRA_AVALON_DEFINITION.id, network)
+export class OptimismLyraAvalonBalanceFetcher implements BalanceFetcher {
+  constructor(
+    @Inject(APP_TOOLKIT) private readonly appToolkit: IAppToolkit,
+    @Inject(LyraAvalonContractFactory) private readonly contractFactory: LyraAvalonContractFactory,
+  ) {}
+
+  async getPoolBalances(address: string) {
+    return await this.appToolkit.helpers.tokenBalanceHelper.getTokenBalances({
+      address,
+      appId,
+      groupId: LYRA_AVALON_DEFINITION.groups.pool.id,
+      network,
+    });
+  }
+
+  async getOptionsBalances(address: string) {
+    const multicall = this.appToolkit.getMulticall(network);
+    const response = await getOptions(this.appToolkit.helpers.theGraphHelper);
+    const markets = await Promise.all(
+      response.markets.map(async market => {
+        const tokenContract = this.contractFactory.optionToken({
+          address: market.optionToken.id.toLowerCase(),
+          network,
+        });
+        const userPositions = await multicall.wrap(tokenContract).getOwnerPositions(address);
+        return {
+          userPositions,
+          marketAddress: market.optionToken.id.toLowerCase(),
+          quoteAddress: market.quoteAddress,
+          strikes: _.flatten(market.boards.map(board => board.strikes)),
+        };
+      }),
+    );
+
+    return this.appToolkit.helpers.contractPositionBalanceHelper.getContractPositionBalances({
+      address,
+      appId,
+      groupId: LYRA_AVALON_DEFINITION.groups.options.id,
+      network,
+      resolveBalances: async ({
+        contractPosition,
+      }: {
+        contractPosition: ContractPosition<LyraAvalonOptionContractPositionDataProps>;
+      }) => {
+        // Find matching market for contract position
+        const market = markets.find(market => market.marketAddress === contractPosition.address);
+        if (!market?.strikes) return [];
+
+        // Extract information from contract position
+        const { strikeId, optionType } = contractPosition.dataProps;
+        const collateralToken = contractPosition.tokens.find(isSupplied)!;
+        const quoteToken = contractPosition.tokens.find(token => token.address === market.quoteAddress.toLowerCase())!;
+
+        // Find matching user position for contract position
+        const userPosition = market.userPositions.find(
+          position => Number(position.strikeId) === strikeId && position.optionType === optionType,
+        );
+        if (!userPosition) return [];
+
+        // Determine price of the contract position strike
+        const strike = market.strikes.find(strike => Number(strike.strikeId) === strikeId);
+        if (!strike) return [];
+        const price = (OPTION_TYPES[optionType].includes('Call') ? strike.callOption : strike.putOption)
```

Comment:
You can put the option price on the data props as well. Try to put all the relevant information in the data props so you don't have to make a GQL request in the balance fetcher (TheGraph is often a lot slower than hitting our RPC nodes in production, so eliminating a GQL request here would make the balance fetcher really snappy)

Reply:
You can even get the market addresses via:

```ts
const contractPositions = await this.appToolkit.getAppContractPositions<LyraAvalonOptionContractPositionDataProps>({
  network,
  appId,
  groupIds: [groupId],
});

const optionAddresses = _.uniq(contractPositions.flatMap(v => v.address));
const userPositions = await Promise.all(optionAddresses.map(optionAddress => {
  const optionContract = this.contractFactory.optionToken({ address: optionAddress, network });
  return multicall.wrap(optionContract).getOwnerPositions(address));
});
```

Reply:
you want this in a separate pr? i see this one is merged already

Reply:
https://github.com/Zapper-fi/studio/pull/529